ytypes/leaf.go

```diff
@@ -508,117 +509,203 @@ the resulting Bgp_Neighbor_RouteReflector would have field
+	ee := ec[1].Interface()
+	if ee != nil {
+		return fmt.Errorf("unmarshaled %v type %T does not have a union type: %v", v, v, ee)
+	}
+
+	util.DbgPrint("unmarshaling %v into type %s", v, reflect.TypeOf(ei))
+
+	eiv := reflect.ValueOf(ei)
+	if destV.Type().Kind() == reflect.Slice {
+		destV.Set(reflect.Append(destV, eiv))
+	} else {
+		destV.Set(eiv)
+	}
+
+	return nil
+}
+
+// getUnionKindsAndTypes returns the YANG kinds of basic types and and 
+// reflect.Types of enum types that are possible for values for the union with 
+// the given schema and type.
+func getUnionKindsAndTypes(schema *yang.Entry, parentT reflect.Type) ([]yang.TypeKind, []reflect.Type, error) {
+	uks, err := getUnionKindsNotEnums(schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	uts, err := schemaToEnumTypes(schema, parentT)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return uks, uts, nil
+}
+
+// getUnionKindsNotEnums returns all the YANG kinds under the given schema node,
+// dereferencing any refs.
+func getUnionKindsNotEnums(schema *yang.Entry) ([]yang.TypeKind, error) {
+	var uks []yang.TypeKind
 	m := make(map[yang.TypeKind]interface{})
-	yts := getUnionTypes(t)
-	for _, yt := range yts {
+	uts, err := getUnionTypesNotEnums(schema, schema.Type)
+	if err != nil {
+		return nil, err
+	}
+	for _, yt := range uts {
 		m[yt.Kind] = nil
 	}
 	for k := range m {
-		out = append(out, k)
+		uks = append(uks, k)
 	}
-	return out
+	return uks, nil
 }
 
-func getUnionTypes(t *yang.YangType) []*yang.YangType {
-	var out []*yang.YangType
-	if t.Kind != yang.Yunion {
-		return []*yang.YangType{t}
-	}
-	for _, t := range t.Type {
-		out = append(out, getUnionTypes(t)...)
+// getUnionTypesNotEnums returns all the non-enum YANG types under the given schema 
+// node, dereferencing any refs.
+func getUnionTypesNotEnums(schema *yang.Entry, yt *yang.YangType) ([]*yang.YangType, error) {
+	var uts []*yang.YangType
+	switch yt.Kind {
+	case yang.Yenum, yang.Yidentityref:
+		// Enum types handled separately.
+		return nil, nil
+	case yang.Yleafref:
+		ns, err := findLeafRefSchema(schema, yt.Path)
+		if err != nil {
+			return nil, err
+		}
+		return getUnionTypesNotEnums(ns, ns.Type)
+	case yang.Yunion:
+		for _, t := range yt.Type {
+			nt, err := getUnionTypesNotEnums(schema, t)
+			if err != nil {
+				return nil, err
+			}
+			uts = append(uts, nt...)
+			if err != nil {
+				return nil, err
+			}
+		}
+	default:
+		uts = []*yang.YangType{yt}
 	}
-	return out
+
+	return uts, nil
 }
 
-func getFieldElemType(parent interface{}, fieldName string) (reflect.Type, error) {
-	t := reflect.TypeOf(parent)
-	ft, ok := t.Elem().FieldByName(fieldName)
-	if !ok {
-		return reflect.TypeOf(nil), fmt.Errorf("%s is not a valid field name in %T", fieldName, parent)
+// schemaToEnumTypes returns the actual enum types (rather than the interface
+// type) for a given schema, which must be for an enum type.
+func schemaToEnumTypes(schema *yang.Entry, t reflect.Type) ([]reflect.Type, error) {
```

Comment:
Assume here that `t` must be the struct that contains the enum field? Consider describing this in the comment above.

Reply:
Done.